airbyte-integrations/connectors/source-zuora/source_zuora/source.py

```diff
@@ -0,0 +1,427 @@
+        try:
+            yield from self.get_data(date_slice=request_kwargs, config=self._config)
+        except ZOQLQueryFieldCannotResolve:
+            """
+            The default cursor_field is "updateddate" sometimes it's not supported by certain streams.
+            We need to swith the default cursor field to alternative one, and retry again the whole operation, submit the new job to the server.
+            We also need to save the state in the end of the sync.
+            So this switch is needed as fast and easy way of resolving the cursor_field for streams that support only the "createddate"
+            """
+            self.cursor_field = self.alt_cursor_field  # cursor_field switch
+            yield from self.get_data(date_slice=request_kwargs, config=self._config)  # retry the whole operation
+        except ZOQLQueryCannotProcessObject:
+            pass
+
+    def parse_response(self, response: requests.Response, **kwargs) -> str:
+        yield from response
+
+
+class ZuoraObjectsBase(ZuoraBase):
+
+    cursor_field = "updateddate"
+    alt_cursor_field = "createddate"
+
+    # setting checkpoint interval to the limit of date-slice
+    @property
+    def state_checkpoint_interval(self) -> int:
+        return self.window_in_days
+
+    def get_cursor_from_schema(self, schema: Dict) -> str:
+        """
+        Get the cursor_field from the stream's schema rather that take it from the class attribute
+        If the stream doesn't support 'updateddate', then we use 'createddate'.
+        """
+        return self.cursor_field if self.cursor_field in schema else self.alt_cursor_field
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        schema = list(ZuoraDescribeObject(self.name, config=self._config).read_records(sync_mode=None))
+        return {"properties": {key: d[key] for d in schema for key in d}}
+
+    def as_airbyte_stream(self) -> AirbyteStream:
+        stream = super().as_airbyte_stream()
+        stream.default_cursor_field = [self.get_cursor_from_schema(stream.json_schema["properties"])]
+        return stream
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        return {self.cursor_field: max(latest_record.get(self.cursor_field, ""), current_stream_state.get(self.cursor_field, ""))}
+
+    def query(self, stream_name: str, cursor_field: str, date_slice: Dict) -> str:
+        return f"""
+            select * from {stream_name} where
+            {cursor_field} >= TIMESTAMP '{date_slice.get('start_date')}' and
+            {cursor_field} <= TIMESTAMP '{date_slice.get('end_date')}'
+            """
+
+    @staticmethod
+    def to_datetime_str(date: datetime) -> str:
+        """
+        :: Output example: '2021-07-15 07:45:55 -07:00' FROMAT : "%Y-%m-%d %H:%M:%S %Z"
+        """
+        return f"{date.strftime('%Y-%m-%d %H:%M:%S %Z')}"
+
```

Comment:
can we return a function result without addl formatting?

Reply:
No, this method formats the input `datetime` into the query-needed formatted date-time string. This is must have.